### ISSUE TITLE
Catch errors mid-chain?

### ISSUE THREAD

**Issue by olsonpm** [petkaantonov/bluebird#476]
### tl;dr
Is catching errors in the middle of  a promise chain considered bad-practice?  Or is there a good reason for it?

### Explanation  

so I just tried the following chain
```
doSomething()
  .then(function() {
    throw new Error('contrived');
  })
  .catch(function(err) {
    console.log('caught');
  })
  .then(function() {
    console.log('got here');
  });
```
And was a little confused that 'got here' gets printed.  I guess the documentation makes this clear:
```
Any exception happening in a .then-chain will propagate to nearest .catch handler.
```
And the fact that catch is an alias for then(null, handler) makes a little more sense of the above behavior.  However it still seems far from the conventional way seen in other OO languages that I've used so I was wondering whether this was on purpose or just a side-effect that should be avoided from an idiomatic perspective.

**Comment by bendrucker** [comment]
First, general "how to" questions, especially ones that relate to promises in general and not Bluebird, should be posted on Stack Overflow. 

It's not a question of best practices. It's about creating the behavior you want. Is an error recoverable via something like a retry? Or do you just need to log it and then move on normally? Then catch it where appropriate and move on. If you're string together a long set of required operations where no single one is optional? Then don't handle errors until the end.

**Comment by olsonpm** [comment]
I guess I should have checked first - but it was my understanding that the A+ promises spec was too general and that my question was specific to an implementation, that being bluebird.

And although you can certainly structure promises depending on the behavior you want - my question related more-so toward the effort taken toward keywords "catch" and "finally" which are both obviously used to familiarize devs with similar concepts.  To me this is the part that's confusing because although they are similar, they can be used with fairly different meaning.

I will post this on stack overflow like you suggest

**Comment by petkaantonov** [comment]
If you wrote the equivalent in synchronous code (Java):

```java
try {
    throw new Exception("caught");
} catch (Exception e) {
    System.out.println("caught");
}
System.out.println("got here");
```

Why is it surprising that "got here" is printed? I am curious because I am writing a beginner's guide and this seems to pop up time to time no matter how much we tell people that it works just like normal code with try catch finally.

**Comment by benjamingr** [comment]
@petkaantonov it's definitely not obvious to newcomers. I've answered at least 5 of these in StackOverflow

**Comment by petkaantonov** [comment]
I think you misunderstood, I said it's obvious in the synchronous code. What I don't understand is why it's not as obvious in the asynchronous code so I can address it in the article.

**Comment by benjamingr** [comment]
Oh, I was just agreeing with you that this should be addressed :D

**Comment by olsonpm** [comment]
@petkaantonov - I think the synchronous equivalent to my example is more like:
```
throw new Exception("contrived");
catch (Exception e) {
  System.out.println("cought");
}
System.out.println("got here");
```
I thought about it some more last night and really the only source of confusion is the lack of a 'try' which strictly defines a block of code to be tried.  Inside a promise chain a 'try' is implied as being any portion of the chain delimited by catch statements.

The following pseudo-code would make it more clear, but also doesn't add any more functionality.
```
Promise.resolve()
  .try(function() {
    return doSomething()
      .then(function() { throw new Error("contrived"); });
  })
  .catch(function(err) {
    console.log('cought');
  })
  .finally(function() {
    console.log('caught');
  });
```
Honestly I kind of like the above because of how it relates to synchronous error handling.  In my mind, a finally statement would only be allowed after a catch, and a catch would only be allowed after a try.  A try would be allowed at any point along a promise chain.

**Comment by petkaantonov** [comment]
> Honestly I kind of like the above because of how it relates to synchronous error handling. In my mind, a finally statement would only be allowed after a try or catch, and a catch would only be allowed after a try. A try would be allowed at any point along a promise chain.

Each `.catch` and `.finally` create an implicit `.try` over the full chain, so this code:

```js
doSomething()
    .then(function() {
        throw new Error("contrived");
    });
    .catch(function(err) {
        console.log('cought');
    })
    .finally(function() {
        console.log('always');
    });
```

Is equivalent to this:

```js
try {
    try {
        doSomething();
        throw new Error("contrived");
    } catch (e) {
        console.log('cought');
    }
} finally {
    console.log('always');
}
```

So if you swapped the position of the `.catch()` and `.finally()` calls in the promise chain, it would still have synchronous equivalent:

```js
try {
    try {
        doSomething();
        throw new Error("contrived");
    } finally {
        console.log('always');
    }
} catch(e) {
    console.log('cought');
}
```

Does this help your understanding?

**Comment by olsonpm** [comment]
> Each .catch and .finally create an implicit .try over the full chain

If that were true, then the following chain syntax wouldn't print 'got here':

```
Promise.resolve()
    .then(function() {
        throw new Error();
    })
    .catch(function() {
        console.log('catch 1');
    })
    .then(function() {
        console.log('got here');
    })
    .then(function() {
        throw new Error();
    })
    .catch(function() {
        console.log('catch 2');
    });
```
So what functionally happens (haven't looked into the code, only observed test behavior) is an implicit try between each catch, which I believe allows for some confusing code.  I just don't expect to see catches and finally's in the middle of a promise chain and I expect other people with more OO backgrounds to feel similarly (though maybe it's just me).

**Comment by petkaantonov** [comment]
Well let's map it to synchronus code according to the rules I gave and run it:

```js
try {
    try {
        throw new Error()
    } catch(e) {
        console.log("catch 1");
    }
    console.log("got here");
    throw new Error();
} catch (e) {
    console.log("catch 2");
}
```

Logs 

    fiddle.jshell.net/:26 catch 1
    fiddle.jshell.net/:28 got here
    fiddle.jshell.net/:31 catch 2
